File: life-assisitant-api/internal/controller/habit.go
// Package controller 习惯控制器
// 规范：controller 只做参数解析 + 调 service + 返回响应
// 业务校验在 service 层完成
package controller

import (
	"github.com/gogf/gf/v2/net/ghttp"

	"github.com/life-assistant/api/internal/consts/ecode"
	"github.com/life-assistant/api/internal/model/dto"
	"github.com/life-assistant/api/internal/response"
	"github.com/life-assistant/api/internal/service"
)

// HabitController 习惯控制器
type HabitController struct{}

var Habit = &HabitController{}

// List GET /api/v1/habits
// Query 参数：status / page / page_size
func (c *HabitController) List(r *ghttp.Request) {
	var req dto.ListHabitsReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Habit().List(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	page := req.Page
	if page <= 0 {
		page = 1
	}
	pageSize := req.PageSize
	if pageSize <= 0 {
		pageSize = 20
	}
	response.Page(r, out.Items, out.Total, page, pageSize)
}

// GetToday GET /api/v1/habits/today
func (c *HabitController) GetToday(r *ghttp.Request) {
	var req dto.GetTodayHabitsReq
	_ = r.Parse(&req) // 可选参数
	out, err := service.Habit().GetToday(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// Get GET /api/v1/habits/:id
func (c *HabitController) Get(r *ghttp.Request) {
	id := r.Get("id").String()
	if id == "" {
		response.Error(r, ecode.MissingRequiredField)
		return
	}
	out, err := service.Habit().Get(r.Context(), id)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// Create POST /api/v1/habits
func (c *HabitController) Create(r *ghttp.Request) {
	var req dto.CreateHabitReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Habit().Create(r.Context(), &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Created(r, out)
}

// Update PATCH /api/v1/habits/:id
func (c *HabitController) Update(r *ghttp.Request) {
	id := r.Get("id").String()
	if id == "" {
		response.Error(r, ecode.MissingRequiredField)
		return
	}
	var req dto.UpdateHabitReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Habit().Update(r.Context(), id, &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

// Delete DELETE /api/v1/habits/:id
func (c *HabitController) Delete(r *ghttp.Request) {
	id := r.Get("id").String()
	if id == "" {
		response.Error(r, ecode.MissingRequiredField)
		return
	}
	if err := service.Habit().Delete(r.Context(), id); err != nil {
		writeError(r, err)
		return
	}
	response.NoContent(r)
}

// Log POST /api/v1/habits/:id/log
func (c *HabitController) Log(r *ghttp.Request) {
	id := r.Get("id").String()
	if id == "" {
		response.Error(r, ecode.MissingRequiredField)
		return
	}
	var req dto.LogHabitReq
	if err := r.Parse(&req); err != nil {
		response.Error(r, ecode.ValidationFailed, err)
		return
	}
	out, err := service.Habit().Log(r.Context(), id, &req)
	if err != nil {
		writeError(r, err)
		return
	}
	response.Success(r, out)
}

File: life-assisitant-ui-desktop/src/api/habit.ts
import request from './request'
import type {
  Habit,
  CreateHabitReq,
  UpdateHabitReq,
  ListHabitsQuery,
  ListHabitsResp,
  LogHabitReq,
} from './types'

/**
 * 习惯 API
 * 规范：md/spec/12-记录.md · 8 数据结构
 * - 列表：GET /habits?status=&date=
 * - 今日：GET /habits/today?date=（含 today_count / today_completed）
 * - 创建：POST /habits
 * - 更新：PATCH /habits/:id
 * - 打卡：POST /habits/:id/log
 * - 删除：DELETE /habits/:id
 */
export const habitApi = {
  /** 获取习惯列表（不含今日打卡数据） */
  list(params?: ListHabitsQuery) {
    return request.get<unknown, ListHabitsResp>('/habits', { params })
  },

  /** 获取今日习惯（含 today_count / today_completed） */
  getToday(params?: { date?: string }) {
    return request.get<unknown, { date: string; items: Habit[]; total: number; done_count: number }>('/habits/today', { params })
  },

  /** 获取单个习惯 */
  get(id: string) {
    return request.get<unknown, Habit>(`/habits/${id}`)
  },

  /** 创建习惯 */
  create(data: CreateHabitReq) {
    return request.post<unknown, Habit>('/habits', data)
  },

  /** 更新习惯 */
  update(id: string, data: UpdateHabitReq) {
    return request.patch<unknown, Habit>(`/habits/${id}`, data)
  },

  /** 删除（软删除 — 归档） */
  remove(id: string) {
    return request.delete(`/habits/${id}`)
  },

  /**
   * 打卡（累加 +1，可重复调用以累加）
   * - 后端回写 updated habit（含最新 today_count / today_done / streak）
   */
  log(id: string, data: LogHabitReq) {
    return request.post<unknown, Habit>(`/habits/${id}/log`, data)
  },
}
